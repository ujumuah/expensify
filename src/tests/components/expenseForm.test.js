 import React from 'React'
 import ExpenseForm from '../../components/pages/expenseForm'
import {shallow} from 'enzyme'
import  data from '../expenses/samples'
import moment from 'moment'
test('Should render form correctly',()=>{
 const wrapper = shallow(<ExpenseForm/>)    
expect(wrapper).toMatchSnapshot()
})
test('should render form with  data correctly',()=>{
    const wrapper = shallow(<ExpenseForm expense={data[0]}/>)
    expect(wrapper).toMatchSnapshot()
})

// stmulating in jest 101
// 1 render the form with shallow render
//  find the  form tag with find() method 
// add the method simulate  and the args ==  the event of ones choice
test('Should render error message when input data is invalid',()=>{
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('form').simulate('submit',{
        preventDefault:()=>{ }
    })
   expect(wrapper.state('error').length).toBeGreaterThan(0)
 })
test('Should align  input field  with the description state',()=>{
    const value="Coffe";
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('input').at(0).simulate('change',{
        target:{value}
    })
     expect(wrapper.state('description')).toBe(value)
})
test('Should align  textarea with the note state',()=>{
    const value="text area sample";
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('textarea').simulate('change',{
        target:{value}
    })
     expect(wrapper.state('notes')).toBe(value)
})
test('Should align input with  the amount state:undefined||invalid data',()=>{
     const wrapper  = shallow(<ExpenseForm/>)
     wrapper.find('input').at(1).simulate('change',{
         target:{value:undefined}
     })
     expect(wrapper.state('input')).toBe(undefined)
})
test('Should align input with  the amount state:data',()=>{
    const value='19.25';
    const wrapper  = shallow(<ExpenseForm/>)
    wrapper.find('input').at(1).simulate('change',{
        target:{value}
    })
    expect(wrapper.state('amount')).toBe(value)
})
test('Form shoould submit  input data to  props',()=>{
    const onSumbitSpy = jest.fn()
 const wrapper = shallow(<ExpenseForm expense={data[1]} FormSubmit={onSumbitSpy}/>)
 expect(wrapper.state('error')).toBe('')
  wrapper.find('form').simulate('submit',{
    preventDefault:()=>{}
  })
   expect(onSumbitSpy).toHaveBeenLastCalledWith({
      description:data[1].description,
      amount:data[1].amount,
      note:data[1].note,
      createAt:data[1].createAt 
  })
})
test('Should check if create key in State is  equal to the moment object',()=>{
    const now =moment(0)
    const wrapper  = shallow(<ExpenseForm/>)
    wrapper.find('SingleDatePicker').prop('onDateChange')(now)
    expect(wrapper.state('createdAt')).toEqual(now)
})
test('checking if the calender Value is Exact or syn with calenderFocuse',()=>{
    const focused = true;
  const  wrapper = shallow(<ExpenseForm/>);
    wrapper.find('SingleDatePicker').prop('onFocusChange')({focused})
    expect(wrapper.state('calenderFocused')).toBe(focused) 

})