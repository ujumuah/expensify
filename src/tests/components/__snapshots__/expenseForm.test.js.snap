// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render form correctly 1`] = `
<div>
  <Component />
  <form
    onSubmit={[Function]}
  >
    <label>
      Description:
      <input
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required="required"
        type="text"
        value=""
      />
    </label>
    <label>
      Amount:
      <input
        autoFocus={true}
        name="Amount"
        onChange={[Function]}
        placeholder="Amount"
        required="required"
        type="text"
        value=""
      />
    </label>
    <SingleDatePicker
      anchorDirection="left"
      customCloseIcon={null}
      customInputIcon={null}
      date={"2021-09-24T13:27:48.581Z"}
      daySize={39}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month",
          "jumpToPrevMonth": "Move backward to switch to the previous month",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
          "moveFocusByOneMonth": "Switch months",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
          "openThisPanel": "Open this panel",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field",
          "selectFocusedDate": "Select the date in focus",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      renderCalendarInfo={null}
      renderDay={null}
      renderMonth={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Explain a litte about the expenses"
      required="required"
      value=""
    />
    <input
      type="submit"
      value="Add expense"
    />
  </form>
</div>
`;

exports[`should render form with  data correctly 1`] = `
<div>
  <Component />
  <form
    onSubmit={[Function]}
  >
    <label>
      Description:
      <input
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required="required"
        type="text"
        value="rent"
      />
    </label>
    <label>
      Amount:
      <input
        autoFocus={true}
        name="Amount"
        onChange={[Function]}
        placeholder="Amount"
        required="required"
        type="text"
        value={1225}
      />
    </label>
    <SingleDatePicker
      anchorDirection="left"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      daySize={39}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month",
          "jumpToPrevMonth": "Move backward to switch to the previous month",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
          "moveFocusByOneMonth": "Switch months",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
          "openThisPanel": "Open this panel",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field",
          "selectFocusedDate": "Select the date in focus",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      renderCalendarInfo={null}
      renderDay={null}
      renderMonth={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Explain a litte about the expenses"
      required="required"
      value="need to pay december rent"
    />
    <input
      type="submit"
      value="Add expense"
    />
  </form>
</div>
`;
